README.MD
# Credits
Thanks to https://github.com/otti/Growatt_ShineWiFi-S

# Growatt_ShineWiFi
Firmware replacement for Growatt ShineWiFi-S (serial), ShineWiFi-X (USB) or custom build sticks (ESP8266/ESP32).
I only use it for the ShineWiFi-X (USB) and Growatt SPH 8000TL3 BH-UP


# How to install
* Checkout this repo
* Setup the IDE of your choice
    * **Recommended:** For platformio just open the project folder and choose the correct env for your hardware  - Use it in Visual Studio Code
* Modify the config.h

After you obtained an image you want to flash:


* Setup  the Stick
** Bridge  GPIO-0 and  GND on the Switch
** Power on  the Stick  with an USB-Cable connected to the Computer - should be detected as COM-Port
* Flash to the Original ShineWiFi-S Stick - Use  "esptool" from python to flash
** Earse all Data: python -m esptool erase_data
** Flash the Firmeware: python -m esptool write_flash --flash_mode dio --flash_size detect 0x0 firmware.bin
* Connect to the setup wifi called GrowattConfig (PW: growsolar) and configure the firmware via the webinterface at http://192.168.4.1
* If you need to reconfigure the stick later on you have to either press the ap button (configured in Config.h) or reset the stick twice within 10sec

## Features
Implemented Features:
* Built-in simple Webserver
* The inverter is queried using Modbus Protocol
* The data received will be transmitted by MQTT to a server of your choice.
* The data received is also provied as JSON
* Show a simple live graph visualization  (`http://<ip>`) with help from highcharts.com
* It supports convenient OTA firmware update (`http://<ip>/firmware`)
* It supports basic access to arbitrary modbus data
* It tries to autodected which stick type to use
* Wifi manager with own access point for initial configuration of Wifi and MQTT server (IP: 192.168.4.1, : GrowattConfig, Pass: growsolar)
* Currently Growatt v1.20, v1.24 and v3.05 protocols are implemented and can be easily extended/changed to fit anyone's needs
* TLS support for esp32

Not supported:
* It does not make use the RTC or SPI Flash of these boards.
* It does not communicate to Growatt Cloud at all.

## Supported sticks/microcontrollers
* ShineWifi-X with a Growatt Inverter connected via USB (USB-Serial Chip from Exar)

## Supported Inverters
* Growatt SPH 4000-10000KTL3 BH (Protocol 124 via Serial)


## Modbus Protocol Versions
The documentation from Growatt on the Modbus interface is avaliable, search for "Growatt PV Inverter Modbus RS485 RTU Protocol" on Google.

The older inverters apparently use Protocol v3.05 from year 2013.
The newer inverters apparently use protocol v1.05 from year 2018.
There is also a new protocol version v1.24 from 2020. (used with SPH4-10KTL3 BH-UP inverter)

## JSON Format Data
For IoT applications the raw data can now read in JSON format (application/json) by calling `http://<ip>/status`

## Homeassistant configuration
This will put the inverter on the energy dashboard.
     
     mqtt:
        sensor:
          - state_topic: "energy/solar"
            unique_id: "growatt_wr_total_production"
            name: "Growatt.TotalGenerateEnergy"
            unit_of_measurement: "kWh"
            value_template: "{{ float(value_json.TotalGenerateEnergy) | round(1) }}"
            device_class: energy
            state_class: total_increasing
            json_attributes_topic: "energy/solar"
            last_reset_topic: "energy/solar"
            last_reset_value_template: "1970-01-01T00:00:00+00:00"
            payload_available: "1"
            availability_mode: latest
            availability_topic: "energy/solar"
            availability_template: "{{ value_json.InverterStatus }}"


To extract the current AC Power you have to add a sensor template.

    template:
      - sensor:
          - name: "Growatt inverter AC Power"
            unit_of_measurement: "W"
            state: "{{ float(state_attr('sensor.growatt_inverter', 'OutputPower')) }}"

## Change log

See [here](CHANGELOG.md)

## Acknowledgements

This arduino sketch will replace the original firmware of the Growatt ShineWiFi stick.

CHANGELOG.md
## 2022-03-12 Fork - Husel2000
* Forked from https://github.com/otti/Growatt_ShineWiFi-S
* Modifyed Growatt124  - remove unused Registers
* Added  Growatt  ModBus RTU Protocol Document

## 2020-10-22 Update
* Added support for ShineWiFi-X (USB). The software will automatically detect on which kind of stick it is running.

* Registers of the inverter can be read or written with an web interface (`http://<ip>/postCommunicationModbus`)

* The total energy is only stored in increments of 0.1 kWh in the inverter. For a smal inverter, this is a large step, especially during winter. 
If the total energy is 0.199 kWh before sunset, the totoal enrgy will be reset to 0.1 kWh next morning. For this reason the "AccumulatedEnergy" field has been implemented. It returns the energy in Wh. It will be zero after every powercycle, but can be set to the value of the previous day over a http request (`http://<ip>/setAccumulatedEnergy`)

* MQTT can be turned off

* The stick can ping an known IP. If there is no answer, the stick will try to reset the WiFi connection. (Disabled by default)

## 2021-10-27 Update
* The automatic detection of the inverter can fail after sunrise. The stick powers up several minutes before the inverter. The detection will only take place directly after power on and will fail because the inverter is not running yet. If the stick can not identify the inverter, it will redo the detection every 2 minutes.
* Debug messages can be read from `<ip>/debug`

* It will automatically detect the stick (USB or serial) and will use the correct baudrate and register set. 
* Added counter for accumulated energy
* Interface for register read/write over web interface added
* MQTT can be turned off by compiler switch
* WiFi connection can be reset if a known IP can not be pinged (optional)

@BeoQ Thanks for your investigations


## 2022-02-18 Update
* Graph will display the data with the timezone of the host pc (UTC was used before)
* Moved the source to a folder which is named like the arduino project, because arduino IDE will otherwise not open the project


## 2022-05-24 Update
* Export already-implemented DcPower and DcInputCurrent values
* Extend RW-Modbus Webpage to also read Input Registers
* Prepare sGrowattData for additional Registers

## 2022-07-31 Update
* Wifi manager added. Thanks to @roel80

## 2022-08-04
* Added support for platform-io (crasu, zinserjan)
* Added support for ESP32 (crasu)

## 2022-08-24 Update
* Redesigned the Growatt class to manage various protocols. (v3.05, v1.20 and v1.24 implemented)
* The new protocol definition allows user to define what modbus register should be read and which ones should be:
	* exported in JSON by calling `http://<ip>/status`
	* displayed on the UI `http://<ip>`
	* added to the graph in the UI
* MQTT_MAX_PACKET_SIZE doesn't need to be updated manually in the library. Instead the `PubSubClient::setBufferSize`has been used to do this dynamically.
* New `http://<ip>/uistatus` has been created to provide data for the UI
* UI changes:
	* Ui is generated dynamically based on the JSON provided.
	* Graph is now able to plot multiple values

## 2023-01-20 Update
* Added tls support for esp32


SRC/ShineWiFi-ModBus/Growatt124.cpp
#include "Arduino.h"

#include "Growatt124.h"

// NOTE: my inverter (SPH4-10KTL3 BH-UP) only manages to read 64 registers in
// one read!
void init_growatt124(sProtocolDefinition_t &Protocol) {
  // definition of input registers
  Protocol.InputRegisterCount = 20;
  // address, value, size, name, multiplier, unit, frontend, plot
  // FEAGMENT 1: BEGIN
  Protocol.InputRegisters[P124_I_STATUS]            = sGrowattModbusReg_t{   0, 0, SIZE_16BIT, "InverterStatus", 1, NONE, true, false};  // #8
  Protocol.InputRegisters[P124_PV1_POWER]           = sGrowattModbusReg_t{   5, 0, SIZE_32BIT, "PowerPV1", 0.1, POWER_W, true, true};  // #5
  Protocol.InputRegisters[P124_PV2_POWER]           = sGrowattModbusReg_t{   9, 0, SIZE_32BIT, "PowerPV2", 0.1, POWER_W, true, true};  // #8
  Protocol.InputRegisters[P124_EPV1_TODAY]          = sGrowattModbusReg_t{  59, 0, SIZE_32BIT, "EnergyPV1Today", 0.1, POWER_KWH, true, false};  // #23
  Protocol.InputRegisters[P124_EPV1_TOTAL]          = sGrowattModbusReg_t{  61, 0, SIZE_32BIT, "EnergyPV1Total", 0.1, POWER_KWH, true, false};  // #24
  Protocol.InputRegisters[P124_EPV2_TODAY]          = sGrowattModbusReg_t{  63, 0, SIZE_32BIT, "EnergyPV2Today", 0.1, POWER_KWH, true, false};  // #26
  Protocol.InputRegisters[P124_EPV2_TOTAL]          = sGrowattModbusReg_t{  65, 0, SIZE_32BIT, "EnergyPV2Total", 0.1, POWER_KWH, true, false};  // #26
  Protocol.InputRegisters[P124_EPV_TOTAL]           = sGrowattModbusReg_t{  91, 0, SIZE_32BIT, "EnergyPVTotal", 0.1, POWER_KWH, true, false};  // #26
  Protocol.InputRegisters[P124_TEMP1]               = sGrowattModbusReg_t{  93, 0, SIZE_16BIT, "InverterTemperature",0.1, TEMPERATURE, true, false};  // #28
  Protocol.InputRegisters[P124_PDISCHARGE]          = sGrowattModbusReg_t{1009, 0, SIZE_32BIT, "PowerBatteryDischarge", 0.1, POWER_W, true, true};  // #31
  Protocol.InputRegisters[P124_PCHARGE]             = sGrowattModbusReg_t{1011, 0, SIZE_32BIT, "PowerBatteryCharge", 0.1, POWER_W, true, true};  // #32  
  Protocol.InputRegisters[P124_SOC]                 = sGrowattModbusReg_t{1014, 0, SIZE_16BIT, "BatteryPercent", 1, PRECENTAGE, true, false};  // #34
  Protocol.InputRegisters[P124_PAC_TO_USER_TOTAL]   = sGrowattModbusReg_t{1021, 0, SIZE_32BIT, "PowerGridImport", 0.1,  POWER_W, true, true};  // #36
  Protocol.InputRegisters[P124_PAC_TO_GRID_TOTAL]   = sGrowattModbusReg_t{1029, 0, SIZE_32BIT, "PowerGridExport",0.1,  POWER_W, true, true};  // #38
  Protocol.InputRegisters[P124_PLOCAL_LOAD_TOTAL]   = sGrowattModbusReg_t{1037, 0, SIZE_32BIT, "PowerHouseCurrent", 0.1,  POWER_W, true, true};  // #40 
  Protocol.InputRegisters[P124_BATTERY_TEMPERATURE] = sGrowattModbusReg_t{1040, 0, SIZE_16BIT, "BatteryTemperature", 0.1,  TEMPERATURE, true, false};  // #41
  Protocol.InputRegisters[P124_ETOGRID_TOTAL]       = sGrowattModbusReg_t{1050, 0, SIZE_32BIT, "EnergyGridExportTotal", 0.1,  POWER_KWH, true, false};  // #46
  Protocol.InputRegisters[P124_ETOUSER_TOTAL]       = sGrowattModbusReg_t{1046, 0, SIZE_32BIT, "EnergyGridImportTotal",0.1,  POWER_KWH, true, false};  // #44
  Protocol.InputRegisters[P124_EDISCHARGE_TOTAL]    = sGrowattModbusReg_t{1054, 0, SIZE_32BIT, "EnergyBatteryDischargeTotal", 0.1,  POWER_KWH, true, false};  // #48
  Protocol.InputRegisters[P124_ECHARGE_TOTAL]       = sGrowattModbusReg_t{1058, 0, SIZE_32BIT, "EnergyBatteryChargeTotal", 0.1,  POWER_KWH, true, false};  // #50
  Protocol.InputRegisters[P124_ETOLOCALLOAD_TOTAL]  = sGrowattModbusReg_t{1062, 0, SIZE_32BIT, "EnergyHouseTotal", 0.1,  POWER_KWH, true, false};  // #52

  Protocol.InputFragmentCount = 3;
  Protocol.InputReadFragments[0] = sGrowattReadFragment_t{0, 50};
  Protocol.InputReadFragments[1] = sGrowattReadFragment_t{53, 43};
  Protocol.InputReadFragments[2] = sGrowattReadFragment_t{1009, 55};

  Protocol.HoldingRegisterCount = 0;
  Protocol.HoldingFragmentCount = 0;
}

SRC/ShineWiFi-ModBus/Growatt.cpp
#include <ModbusMaster.h>
#include <ArduinoJson.h>

#include "GrowattTypes.h"
#include "Growatt.h"
#include "Config.h"
#ifndef _SHINE_CONFIG_H_
#error Please rename Config.h.example to Config.h
#endif

#if GROWATT_MODBUS_VERSION == 120
#include "Growatt120.h"
#elif GROWATT_MODBUS_VERSION == 124
#include "Growatt124.h"
#elif GROWATT_MODBUS_VERSION == 305
#include "Growatt305.h"
#elif GROWATT_MODBUS_VERSION == 5000
#include "GrowattSPF.h"
#else
#error "Unsupported Growatt Modbus version"
#endif

ModbusMaster Modbus;

// Constructor
Growatt::Growatt() {
  _eDevice = Undef_stick;
  _PacketCnt = 0;
}

void Growatt::InitProtocol() {
/**
 * @brief Initialize the protocol struct
 * @param version The version of the modbus protocol to use
 */
#if GROWATT_MODBUS_VERSION == 120
  init_growatt120(_Protocol);
#elif GROWATT_MODBUS_VERSION == 124
  init_growatt124(_Protocol);
#elif GROWATT_MODBUS_VERSION == 305
  init_growatt305(_Protocol);
#elif GROWATT_MODBUS_VERSION == 5000
  init_growattSPF(_Protocol);
#else
#error "Unsupported Growatt Modbus version"
#endif
}

void Growatt::begin(Stream& serial) {
  /**
   * @brief Set up communication with the inverter
   * @param serial The serial interface
   */
  uint8_t res;

#if SIMULATE_INVERTER == 1
  _eDevice = SIMULATE_DEVICE;
#else
  // init communication with the inverter
  Serial.begin(9600);
  Modbus.begin(1, serial);
  res = Modbus.readInputRegisters(0, 1);
  if (res == Modbus.ku8MBSuccess) {
    _eDevice = ShineWiFi_S;  // Serial
  } else {
    delay(1000);
    Serial.begin(115200);
    Modbus.begin(1, serial);
    res = Modbus.readInputRegisters(0, 1);
    if (res == Modbus.ku8MBSuccess) {
      _eDevice = ShineWiFi_X;  // USB
    }
    delay(1000);
  }
#endif
}

eDevice_t Growatt::GetWiFiStickType() {
  /**
   * @brief After initialisation the type of the wifi stick is known
   * @returns eDevice_t type of the wifi stick
   */

  return _eDevice;
}

bool Growatt::ReadInputRegisters() {
  /**
   * @brief Read the input registers from the inverter
   * @returns true if data was read successfully, false otherwise
   */
  uint16_t registerAddress;
  uint8_t res;

  // read each fragment separately
  for (int i = 0; i < _Protocol.InputFragmentCount; i++) {
    res = Modbus.readInputRegisters(_Protocol.InputReadFragments[i].StartAddress, _Protocol.InputReadFragments[i].FragmentSize);
    if (res == Modbus.ku8MBSuccess) {
      for (int j = 0; j < _Protocol.InputRegisterCount; j++) {
        // make sure the register we try to read is in the fragment
        if (_Protocol.InputRegisters[j].address >= _Protocol.InputReadFragments[i].StartAddress) {
          // when we exceed the fragment size, skip to new fragment
          if (_Protocol.InputRegisters[j].address >= _Protocol.InputReadFragments[i].StartAddress + _Protocol.InputReadFragments[i].FragmentSize) {
            break;
          }
          // let's say the register address is 1013 and read window is 1000-1050
          // that means the response in the buffer is on position 1013 - 1000 =
          // 13
          registerAddress = _Protocol.InputRegisters[j].address - _Protocol.InputReadFragments[i].StartAddress;
          if (_Protocol.InputRegisters[j].size == SIZE_16BIT || _Protocol.InputRegisters[j].size == SIZE_16BIT_S) {
            _Protocol.InputRegisters[j].value = Modbus.getResponseBuffer(registerAddress);
          } else {
            _Protocol.InputRegisters[j].value = (Modbus.getResponseBuffer(registerAddress) << 16) + Modbus.getResponseBuffer(registerAddress + 1);
          }
        }
      }
    } else {
      return false;
    }
  }
  return true;
}

bool Growatt::ReadHoldingRegisters() {
  /**
   * @brief Read the holding registers from the inverter
   * @returns true if data was read successfully, false otherwise
   */
  uint16_t registerAddress;
  uint8_t res;

  // read each fragment separately
  for (int i = 0; i < _Protocol.HoldingFragmentCount; i++) {
    res = Modbus.readHoldingRegisters(
        _Protocol.HoldingReadFragments[i].StartAddress,
        _Protocol.HoldingReadFragments[i].FragmentSize);
    if (res == Modbus.ku8MBSuccess) {
      for (int j = 0; j < _Protocol.HoldingRegisterCount; j++) {
        if (_Protocol.HoldingRegisters[j].address >=
            _Protocol.HoldingReadFragments[i].StartAddress) {
          if (_Protocol.HoldingRegisters[j].address >=
              _Protocol.HoldingReadFragments[i].StartAddress +
                  _Protocol.HoldingReadFragments[i].FragmentSize)
            break;
          registerAddress = _Protocol.HoldingRegisters[j].address -
                            _Protocol.HoldingReadFragments[i].StartAddress;
          if (_Protocol.HoldingRegisters[j].size == SIZE_16BIT ||
              _Protocol.HoldingRegisters[j].size == SIZE_16BIT_S) {
            _Protocol.HoldingRegisters[j].value =
                Modbus.getResponseBuffer(registerAddress);
          } else {
            _Protocol.HoldingRegisters[j].value =
                (Modbus.getResponseBuffer(registerAddress) << 16) +
                Modbus.getResponseBuffer(registerAddress + 1);
          }
        }
      }
    } else {
      return false;
    }
  }
  return true;
}

bool Growatt::ReadData() {
  /**
   * @brief Reads the data from the inverter and updates the internal data
   * structures
   * @returns true if data was read successfully, false otherwise
   */

  _PacketCnt++;
  _GotData = ReadInputRegisters() && ReadHoldingRegisters();
  return _GotData;
}

sGrowattModbusReg_t Growatt::GetInputRegister(uint16_t reg) {
  /**
   * @brief get the internal representation of the input register
   * @param reg the register to get
   * @returns the register value
   */
  if (_GotData == false) {
    ReadData();
  }
  return _Protocol.InputRegisters[reg];
}

sGrowattModbusReg_t Growatt::GetHoldingRegister(uint16_t reg) {
  /**
   * @brief get the internal representation of the holding register
   * @param reg the register to get
   * @returns the register value
   */
  if (_GotData == false) {
    ReadData();
  }
  return _Protocol.HoldingRegisters[reg];
}

bool Growatt::ReadHoldingReg(uint16_t adr, uint16_t* result) {
  /**
   * @brief read 16b holding register
   * @param adr address of the register
   * @param result pointer to the result
   * @returns true if successful
   */
  uint8_t res = Modbus.readHoldingRegisters(adr, 1);
  if (res == Modbus.ku8MBSuccess) {
    *result = Modbus.getResponseBuffer(0);
    return true;
  }
  return false;
}

bool Growatt::ReadHoldingReg(uint16_t adr, uint32_t* result) {
  /**
   * @brief read 32b holding register
   * @param adr address of the register
   * @param result pointer to the result
   * @returns true if successful
   */
  uint8_t res = Modbus.readHoldingRegisters(adr, 2);
  if (res == Modbus.ku8MBSuccess) {
    *result = (Modbus.getResponseBuffer(0) << 16) + Modbus.getResponseBuffer(1);
    return true;
  }
  return false;
}

bool Growatt::WriteHoldingReg(uint16_t adr, uint16_t value) {
  /**
   * @brief write 16b holding register
   * @param adr address of the register
   * @param value value to write to the register
   * @returns true if successful
   */
  uint8_t res = Modbus.writeSingleRegister(adr, value);
  if (res == Modbus.ku8MBSuccess) {
    return true;
  }
  return false;
}

bool Growatt::ReadInputReg(uint16_t adr, uint16_t* result) {
  /**
   * @brief read 16b input register
   * @param adr address of the register
   * @param result pointer to the result
   * @returns true if successful
   */
  uint8_t res = Modbus.readInputRegisters(adr, 1);
  if (res == Modbus.ku8MBSuccess) {
    *result = Modbus.getResponseBuffer(0);
    return true;
  }
  return false;
}

bool Growatt::ReadInputReg(uint16_t adr, uint32_t* result) {
  /**
   * @brief read 32b input register
   * @param adr address of the register
   * @param result pointer to the result
   * @returns true if successful
   */
  uint8_t res = Modbus.readInputRegisters(adr, 2);
  if (res == Modbus.ku8MBSuccess) {
    *result = (Modbus.getResponseBuffer(0) << 16) + Modbus.getResponseBuffer(1);
    return true;
  }
  return false;
}

double Growatt::_round2(double value) {
  return (int)(value * 100 + 0.5) / 100.0;
}

void Growatt::JSONAddReg(sGrowattModbusReg_t* reg, JsonDocument* doc) {
  char* name = reg->name;
  RegisterSize_t size = reg->size;
  float mult = reg->multiplier;
  uint32_t value = reg->value;

  switch (size) {
    case SIZE_16BIT_S:
      (*doc)[name] = (mult == (int)mult) ? (int16_t)value * mult
                                         : _round2((int16_t)value * mult);
      break;
    case SIZE_32BIT_S:
      (*doc)[name] = (mult == (int)mult) ? (int32_t)value * mult
                                         : _round2((int32_t)value * mult);
      break;
    default:
      (*doc)[name] = (mult == (int)mult) ? value * mult : _round2(value * mult);
  }
}

void Growatt::CreateJson(char* Buffer, const char* MacAddress) {
  StaticJsonDocument<2048> doc;

#if SIMULATE_INVERTER != 1
  for (int i = 0; i < _Protocol.InputRegisterCount; i++)
    JSONAddReg(&_Protocol.InputRegisters[i], &doc);

  for (int i = 0; i < _Protocol.HoldingRegisterCount; i++)
    JSONAddReg(&_Protocol.HoldingRegisters[i], &doc);
#else
#warning simulating the inverter
  doc["Status"] = 1;
  doc["DcPower"] = 230;
  doc["DcVoltage"] = 70.5;
  doc["DcInputCurrent"] = 8.5;
  doc["AcFreq"] = 50.00;
  doc["AcVoltage"] = 230.0;
  doc["AcPower"] = 0.00;
  doc["EnergyToday"] = 0.3;
  doc["EnergyTotal"] = 49.1;
  doc["OperatingTime"] = 123456;
  doc["Temperature"] = 21.12;
  doc["AccumulatedEnergy"] = 320;
  doc["EnergyToday"] = 0.3;
  doc["EnergyToday"] = 0.3;
#endif  // SIMULATE_INVERTER
  doc["Mac"] = MacAddress;
  doc["Cnt"] = _PacketCnt;
  serializeJson(doc, Buffer, MQTT_MAX_PACKET_SIZE);
}

void Growatt::CreateUIJson(char* Buffer) {
  StaticJsonDocument<2048> doc;
  const char* unitStr[] = {"", "W", "kWh", "V", "A", "s", "%", "Hz", "°C"};
  const char* statusStr[] = {"(Waiting)", "(Normal Operation)", "State: 2", "(Error)", "State: 4", "State: 5", "State: 6", "EPS"};
  const int statusStrLength = sizeof(statusStr) / sizeof(char*);

#if SIMULATE_INVERTER != 1
  for (int i = 0; i < _Protocol.InputRegisterCount; i++) {
    if (_Protocol.InputRegisters[i].frontend == true || _Protocol.InputRegisters[i].plot == true) {
      JsonArray arr = doc.createNestedArray(_Protocol.InputRegisters[i].name);

      // value
      if (_Protocol.InputRegisters[i].multiplier == (int)_Protocol.InputRegisters[i].multiplier) {
        arr.add(_Protocol.InputRegisters[i].value * _Protocol.InputRegisters[i].multiplier);
      } else {
        arr.add(_round2(_Protocol.InputRegisters[i].value * _Protocol.InputRegisters[i].multiplier));
      }
      if (strcmp(_Protocol.InputRegisters[i].name, "InverterStatus") == 0 &&
          _Protocol.InputRegisters[i].value < statusStrLength) { arr.add(statusStr[_Protocol.InputRegisters[i].value]);  // use unit for status
      } else {
        arr.add(unitStr[_Protocol.InputRegisters[i].unit]);  // unit
      }
      arr.add(_Protocol.InputRegisters[i].plot);  // should be plotted
    }
  }
  for (int i = 0; i < _Protocol.HoldingRegisterCount; i++) {
    if (_Protocol.HoldingRegisters[i].frontend == true || _Protocol.HoldingRegisters[i].plot == true) {
      JsonArray arr = doc.createNestedArray(_Protocol.HoldingRegisters[i].name);

      // value
      if (_Protocol.HoldingRegisters[i].multiplier == (int)_Protocol.HoldingRegisters[i].multiplier) {
        arr.add(_Protocol.HoldingRegisters[i].value * _Protocol.HoldingRegisters[i].multiplier);
      } else {
        arr.add(_round2(_Protocol.HoldingRegisters[i].value * _Protocol.HoldingRegisters[i].multiplier));
      }
      if (strcmp(_Protocol.HoldingRegisters[i].name, "InverterStatus") == 0 && _Protocol.HoldingRegisters[i].value < statusStrLength) {
        arr.add(statusStr[_Protocol.HoldingRegisters[i].value]);  // use unit for status
      } else {
        arr.add(unitStr[_Protocol.HoldingRegisters[i].unit]);  // unit
      }
      arr.add(_Protocol.HoldingRegisters[i].plot);  // should be plotted
    }
  }
#else
#warning simulating the inverter
  JsonArray arr = doc.createNestedArray("Status");
  arr.add(1);
  arr.add("(Normal Operation)");
  arr.add(false);
  arr = doc.createNestedArray("DcPower");
  arr.add(230);
  arr.add("W");
  arr.add(true);
  arr = doc.createNestedArray("DcVoltage");
  arr.add(70.5);
  arr.add("V");
  arr.add(false);
  arr = doc.createNestedArray("DcInputCurrent");
  arr.add(8.5);
  arr.add("A");
  arr.add(false);
  arr = doc.createNestedArray("AcFreq");
  arr.add(50);
  arr.add("Hz");
  arr.add(false);
  arr = doc.createNestedArray("AcVoltage");
  arr.add(230);
  arr.add("V");
  arr.add(false);
  arr = doc.createNestedArray("AcPower");
  arr.add(0.00);
  arr.add("W");
  arr.add(false);
  arr = doc.createNestedArray("EnergyToday");
  arr.add(0.3);
  arr.add("kWh");
  arr.add(false);
  arr = doc.createNestedArray("EnergyTotal");
  arr.add(49.1);
  arr.add("kWh");
  arr.add(false);
  arr = doc.createNestedArray("OperatingTime");
  arr.add(123456);
  arr.add("s");
  arr.add(false);
  arr = doc.createNestedArray("Temperature");
  arr.add(21.12);
  arr.add("C");
  arr.add(false);
  arr = doc.createNestedArray("AccumulatedEnergy");
  arr.add(320);
  arr.add("kWh");
  arr.add(false);
  arr = doc.createNestedArray("EnergyToday");
  arr.add(0.3);
  arr.add("kWh");
  arr.add(false);
#endif  // SIMULATE_INVERTER

  serializeJson(doc, Buffer, 4096);
}
